### PR TITLE
Add jq adapter

### PR DESCRIPTION
**Pull request by odnoletkov**
See https://github.com/stedolan/jq

### CODE REVIEW

File: autoload/db/adapter/jq.vim [hunk 1]
```diff
@@ -0,0 +1,26 @@
...
+endfunction
+
+function! db#adapter#jq#test_file(file) abort
+  if readfile(a:file, '', 1)[0] =~# '^{' && readfile(a:file, '', -1)[0] =~# '}$'
+    return 1
+  endif
+endfunction
+
+function! db#adapter#jq#command(url) abort
+  return 'jq --from-file /dev/stdin ' . db#url#file_path(a:url)
```
Comment (odnoletkov) [inline autoload/db/adapter/jq.vim:17]: Unfortunately parameters order is important here so can't use `input_flag`:

```
$ jq -f <(echo .) <(echo 1)
1
$ jq <(echo 1) -f <(echo .)
parse error: Invalid numeric literal at line 2, column 0
```

**Comment by tpope** [comment]
It's more than a stretch to call `jq` a database, but I can't think of any reason not to support it.

* Make sure `#test_file()` doesn't blow up on an empty file, and correctly handles leading/trailing whitespace (including blank lines?) and files containing `[]` arrays. Error on the side of false positives, not false negatives.
* Without an `#input_extension()` it will default to `sql`, which is wrong. Use `jq`, I guess, unless you have a better idea.
* We probably want an `#output_extension()` of `json`.

**Comment by odnoletkov** [comment]
> * We probably want an `#output_extension()` of `json`.

Done. This enables proper highlighting in the output buffer. Technically jq can output non-valid json, for example a single string – this is highlighted as an error. But I think this is a small price to pay for the convenience of proper json filetype in most cases.

> * Without an `#input_extension()` it will default to `sql`, which is wrong. Use `jq`, I guess, unless you have a better idea.

Done. Although I couldn't find what is affected by this.

> * Make sure `#test_file()` doesn't blow up on an empty file, and correctly handles leading/trailing whitespace (including blank lines?) and files containing `[]` arrays. Error on the side of false positives, not false negatives.

I've made the change to actually invoke jq to validate the input file. It is fast and fails early on large files. It is also more correct because technically jq accepts sequence of valid json values, not just array or object.

But let me know if there are practical downsides to this (security?) – I'll try my best to write a manual `#test_file()` as you described.

---

Also added special case to run jq w/o input (`--null-input`) for empty path.

What should we do with the interactive mode? jq doesn't support it out of the box (but there are [addons](https://github.com/fiatjaf/jiq)). Currently this error is printed for interactive mode command:

    E117: Unknown function: db#adapter#jq#interactive

**Comment by tpope** [comment]
> Technically jq can output non-valid json, for example a single string – this is highlighted as an error. But I think this is a small price to pay for the convenience of proper json filetype in most cases.

Thanks for reminding me how much `syntax/json.vim` sucks.

> I've made the change to actually invoke jq to validate the input file. It is fast and fails early on large files. It is also more correct because technically jq accepts sequence of valid json values, not just array or object.
> 
> But let me know if there are practical downsides to this (security?) – I'll try my best to write a manual `#test_file()` as you described.

The downside is that `system()` is not guaranteed to be fast. It's slow on Windows, but also when people do annoying things like `:set shellcmdflag=-ic`, and it gets called repeatedly to filter out tab completion results.

Now that I think about it I'm not sure we should define it at all. For `:DB /file.sqlite3`, `sqlite3` is the canonical handler. But JSON files, `jq` is a niche tool with a cult following. JSON isn't a `jq` database, it's a general purpose file format.

> What should we do with the interactive mode? jq doesn't support it out of the box (but there are [addons](https://github.com/fiatjaf/jiq)). Currently this error is printed for interactive mode command:
> 
> ```
> E117: Unknown function: db#adapter#jq#interactive
> ```

We should short circuit with error like `DB: interactive mode not supported for <adapter>` if `db#adapter#supports()` returns false. Use the `return 'echoerr...'` pattern.

File: autoload/db/adapter/jq.vim [hunk 2]
```diff
@@ -0,0 +1,33 @@
...
+  let input = db#url#file_path(a:url)
+  if input =~# '^[\/]\=$'
+    let input = '--null-input'
+  endif
+  return 'jq --from-file /dev/stdin ' . input
+endfunction
+
+function! db#adapter#jq#tables(url) abort
+  return split(system('jq --raw-output "[.. | objects | keys[]] | unique[]" ' . db#url#file_path(a:url)), "\n")
+endfunction
```
Comment (tpope) [inline autoload/db/adapter/jq.vim:33]: Needs more `shellescape()`

**Comment by odnoletkov** [comment]
* Remove `#test_file()` to not claim ownership of JSON files
* Fail early if `#interactive()` is not supported
* `shellescape()` paths

**Pull request merged**